File: 프로그래머스/lv2/84512. 모음 사전/모음 사전.java
import java.util.*;

class Solution {
    
    String[] words = {"A", "E", "I", "O", "U"};
    Map<String, Integer> order = new HashMap<>();
    int idx = 1;
    
    public int solution(String word) {
        int answer = 0;
        dfs(0, "");
        answer = order.get(word);
        return answer;
    }
    
    public void dfs(int cnt, String word){
        if(cnt == 5){
            return;
        }
        
        for(int i=0; i<5; i++){
            String curWord = word + words[i];
            //stem.out.println(curWord);
            order.put(curWord, idx);
            idx++;
            dfs(cnt+1, curWord);
            
        }
        
        
    }
}

File: 프로그래머스/2/84512. 모음 사전/모음 사전.py
def solution(word):
    words_list = []
    dfs(0, "", words_list)
    return words_list.index(word)

def dfs(depth, word, words_list):
    
    words_list.append(word)
    
    if depth == 5:
        return
    for c in "AEIOU":
        dfs(depth+1, word+c, words_list)
        